scripts/destroy_stage.ts

```diff
@@ -0,0 +1,239 @@
+        }
+
+        try {
+            // construct the delete
+            const deleteParams: AWS.S3.DeleteObjectsRequest = {
+                Bucket: bucket.PhysicalResourceId ?? '',
+                Delete: {
+                    Objects: keys,
+                },
+            }
+
+            // delete all the files, including their versions
+            const deleteObjectsResponse = await s3
+                .deleteObjects(deleteParams)
+                .promise()
+            console.log(
+                `deleted objects: ${deleteObjectsResponse.$response.data}`
+            )
+            if (deleteObjectsResponse.$response.error != null) {
+                return new Error(
+                    `Error on deleteObjects: ${deleteObjectsResponse.$response.error}`
+                )
+            }
+        } catch (err) {
+            return new Error(
+                `Error clearing bucket: ${bucket.PhysicalResourceId}: ${err}`
+            )
+        }
+    })
+
+    return {}
+}
+
+async function getBucketsInStack(
+    stackName: string
+): Promise<AWS.CloudFormation.StackResource[] | Error> {
+    // get all the resources in the stack
+    try {
+        const stack = await cf
+            .describeStackResources({ StackName: stackName })
+            .promise()
+
+        if (stack.StackResources === undefined) {
+            return new Error('could not find stack')
+        }
+
+        // filter the resources to get our S3 buckets
+        return stack.StackResources.filter((resource) => {
+            return resource.ResourceType === 'AWS::S3::Bucket'
+        })
+    } catch (err) {
+        return new Error(`Could not get stack resources: ${err}`)
+    }
+}
+
+async function turnOffVersioningOnBucket(
+    bucket: AWS.CloudFormation.StackResource
+): Promise<{} | Error> {
+    console.log(
+        `Turning off bucket versioning on bucket: ${bucket.PhysicalResourceId}`
+    )
+
+    const versionParams = {
+        Bucket: bucket.PhysicalResourceId ?? '',
+        VersioningConfiguration: { Status: 'Suspended' },
+    }
+
+    try {
+        await s3.putBucketVersioning(versionParams).promise()
+    } catch (err) {
+        return new Error(`Could not turn off bucket versioning: ${err}`)
+    }
+
+    return {}
+}
+
+async function getVersionedFilesInBucket(
+    bucket: AWS.CloudFormation.StackResource
+): Promise<s3ObjectKey[] | Error> {
+    // Versioned buckets will have extra files in them, all of which
+    // must be cleared out before CloudFormation can delete the bucket.
+    // We have to remove all the versions and any delete markers that exist.
+    const bucketParams = {
+        Bucket: bucket.PhysicalResourceId ?? '',
+    }
+
+    // get all versioned objects
+    let versionKeys: s3ObjectKey[] = []
+    let deleteMarkerKeys: s3ObjectKey[] = []
+
+    const objectVersions = await s3.listObjectVersions(bucketParams).promise()
+    // AWSError is unfortunately not backed by Error, so we can't just
```

Comment:
ahh ok awesome! I had been thinking of writing a type guard around `AWSError` but figured it was too much for just this little destroy script. This looks like it would be helpful, thanks!

Reply:
what is returned when there is an error?

Reply:
Oh, seeing this down below, maybe we want something like this here too?
```
if (stack.$response.error != null) {
         return new Error(
             'Error on describeStacks: ' + stack.$response.error.message
         )
     }
```

Reply:
oh, yup, I can add a check there too -- thanks!